pkg/executables/cmk.go

```diff
@@ -0,0 +1,363 @@
+	offerings, err := c.ListServiceOfferings(ctx, computeOffering)
+	if err != nil {
+		return "", fmt.Errorf("compute offering %s not found. error: %v", computeOffering, err)
+	} else if len(offerings) > 1 {
+		return "", fmt.Errorf("duplicate compute offering %s found", computeOffering)
+	} else if len(offerings) == 0 {
+		return "", fmt.Errorf("compute offering %s not found", computeOffering)
+	}
+
+	return offerings[0].Name, nil
+}
+
+func (c *Cmk) SearchDiskOffering(ctx context.Context, domain string, zone string, account string, diskOffering string) (string, error) {
+	if diskOffering == "" {
+		return diskOffering, nil
+	}
+	offerings, err := c.ListDiskOfferings(ctx, diskOffering)
+	if err != nil {
+		return "", fmt.Errorf("disk offering %s not found. error: %v", diskOffering, err)
+	} else if len(offerings) > 1 {
+		return "", fmt.Errorf("duplicate disk offering %s found", diskOffering)
+	} else if len(offerings) == 0 {
+		return "", fmt.Errorf("disk offering %s not found", diskOffering)
+	}
+	return offerings[0].Name, nil
+}
+
+func (c *Cmk) SearchAffinityGroups(ctx context.Context, domain string, zone string, account string, affinityGroupIds []string) error {
+	for _, affinityGroupId := range affinityGroupIds {
+		affinityGroup, err := c.ListAffinityGroupsById(ctx, affinityGroupId)
+		if err != nil {
+			return fmt.Errorf("affinity group %s not found. error: %v", affinityGroupId, err)
+		} else if len(affinityGroup) > 1 {
+			return fmt.Errorf("duplicate affinity group %s found", affinityGroupId)
+		} else if len(affinityGroup) == 0 {
+			return fmt.Errorf("affinity group %s not found", affinityGroupId)
+		}
+	}
+	return nil
+}
+
+// cmkExecConfig contains transient information for the execution of cmk commands
+// It must be cleaned after each execution to prevent side effects from past executions options
+type cmkExecConfig struct {
+	env        map[string]string
+	ConfigFile string
+}
+
+func NewCmk(executable Executable, writer filewriter.FileWriter) *Cmk {
+	return &Cmk{
+		writer:     writer,
+		executable: executable,
+		retrier:    retrier.NewWithMaxRetries(maxRetriesCmk, backOffPeriodCmk),
+	}
+}
+
+// ValidateCloudStackConnection Calls `cmk sync` to ensure that the endpoint and credentials + domain are valid
+func (c *Cmk) ValidateCloudStackConnection(ctx context.Context) error {
+	buffer, err := c.exec(ctx, "sync")
+	if err != nil {
+		return fmt.Errorf("error validating cloudstack connection for cmk config %s: %v", buffer.String(), err)
+	}
+	logger.MarkPass("Connected to CloudStack server")
+	return nil
+}
+
+// TODO: Add support for domain, account filtering
+func (c *Cmk) ListTemplates(ctx context.Context, template string) ([]types.CmkTemplate, error) {
+	result, err := c.execWithNameAndIdFilters(ctx, template, "list", "templates", "templatefilter=all", "listall=true")
+	if err != nil {
+		return make([]types.CmkTemplate, 0), fmt.Errorf("error getting templates info: %v", err)
```

Comment:
you can just return
```suggestion
		return nil, fmt.Errorf("error getting templates info: %v", err)
```

Reply:
I'll re-evaluate the contract between this library and the business logic where it'll be consumed to determine what to return here in the next revision